src/GafferScene/Unencapsulate.cpp

```diff
@@ -0,0 +1,287 @@
+		}
+	}
+	return capsule->scene()->object( concatScenePath( capsule->root(), branchPath ) );
+}
+
+bool Unencapsulate::affectsBranchChildNames( const Gaffer::Plug *input ) const
+{
+	return input == inPlug()->objectPlug();
+}
+
+void Unencapsulate::hashBranchChildNames( const ScenePath &parentPath, const ScenePath &branchPath, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	IECore::ConstObjectPtr capsuleObject = inPlug()->object( parentPath );
+	const Capsule *capsule = IECore::runTimeCast< const Capsule >( capsuleObject.get() );
+	if( !capsule )
+	{
+		h.append( 0 );
+		return;
+	}
+	if( branchPath.size() == 0 )
+	{
+		h = capsule->scene()->childNamesHash( capsule->root() );
+	}
+	else
+	{
+		h = capsule->scene()->childNamesHash( concatScenePath( capsule->root(), branchPath ) );
+	}
+}
+
+IECore::ConstInternedStringVectorDataPtr Unencapsulate::computeBranchChildNames( const ScenePath &parentPath, const ScenePath &branchPath, const Gaffer::Context *context ) const
+{
+	IECore::ConstObjectPtr capsuleObject = inPlug()->object( parentPath );
+	const Capsule *capsule = IECore::runTimeCast< const Capsule >( capsuleObject.get() );
+	if( !capsule )
+	{
+		return new InternedStringVectorData();
+	}
+
+	if( branchPath.size() == 0 )
+	{
+		return capsule->scene()->childNames( capsule->root() );
+	}
+	else
+	{
+		return capsule->scene()->childNames( concatScenePath( capsule->root(), branchPath ) );
+	}
+}
+
+bool Unencapsulate::affectsBranchSetNames( const Gaffer::Plug *input ) const
+{
+	return input == inPlug()->objectPlug();
+}
+
+void Unencapsulate::hashBranchSetNames( const ScenePath &parentPath, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	h = inPlug()->setNamesPlug()->hash();
+}
+
+IECore::ConstInternedStringVectorDataPtr Unencapsulate::computeBranchSetNames( const ScenePath &parentPath, const Gaffer::Context *context ) const
+{
+	// We have a standard that any seen containing capsules must contain all the sets used in the capsules
+	// in their list of set names, even if those sets are empty until the capsules are expanded
+	return inPlug()->setNamesPlug()->getValue();
+}
+
+bool Unencapsulate::affectsBranchSet( const Gaffer::Plug *input ) const
+{
+	return input == inPlug()->objectPlug();
+}
+
+void Unencapsulate::hashBranchSet( const ScenePath &parentPath, const IECore::InternedString &setName, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	BranchCreator::hashBranchSet( parentPath, setName, context, h );
+
+	IECore::ConstObjectPtr capsuleObject = inPlug()->object( parentPath );
+	const Capsule *capsule = IECore::runTimeCast< const Capsule >( capsuleObject.get() );
+	if( !capsule )
+	{
+		h.append( 0 );
+		return;
+	}
+
+	h.append( capsule->scene()->setPlug()->hash() );
+
+	// The resulting set depends on which capsule we're reading, we could have many different capsules
+	// from the same source scene
+	for( const IECore::InternedString &n : parentPath )
+	{
+		h.append( n );
+	}
```

Comment:
I think there's an array form of append we can use here - `h.append( n.data(), n.size()`, summink like that (I didn't check). But isn't it `capsule->root()` we should be hashing here? If we were to swap one capsule out for another in the same location, the `parentPath` wouldn't change even though the `root()` might. I think we need to match what is used in `computeBranchSet()`, and that's the root.

Reply:
Yeah, you're right.  The set depends on both the source path and the destination path, but BranchCreator::hashSet already does this with the parentPath.